Sources/third_party/Experimental/Complex.swift

```diff
@@ -0,0 +1,339 @@
+// Copyright 2017-2019 Xiaodi Wu and The TensorFlow Authors. All Rights Reserved.
+//
+// Licensed under the MIT License (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://opensource.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//  Note
+//  ====
+//
+//  Code in libc++ is dual-licensed under the MIT and UIUC/NCSA licenses.
+//  Copyright © 2009-2017 contributors to the LLVM/libc++ project.
+/// A type to represent a complex value in Cartesian form.
+///
+/// Create new instances of `Complex<T>` using integer or floating-point
+/// literals and the imaginary unit `Complex<T>.i`. For example:
+///
+/// ```swift
+/// let x: Complex<Double> = 2 + 4 * .i
+/// ```
+///
+/// Additional Considerations
+/// -------------------------
+///
+/// Floating-point types have special values that represent infinity or NaN
+/// ("not a number"). Complex functions in different languages may return
+/// different results when working with special values.
+
+struct Complex<T: FloatingPoint> {
+    var real: T
+    var imaginary: T
```

Comment:
Have you tried making `real` and `imaginary` differentiable? If you do that, then the compiler will in principle be able to automatically compute derivatives for all the complex operations because they just take `real` and `imaginary` parts, then perform differentiable operations on those, then construct a new complex number (which is also a differentiable operation)!

In practice, the compiler can't quite derive these yet because most of the complex operations are implemented in terms of mutation and control flow that the compiler can't handle. But soon it should be able to handle it!

It would be very interesting to add some tests that define a few operations in ways that autodiff can handle, and then see if autodiff computes the expected derivatives for those. e.g.

```
func simplifiedAdd(_ lhs: Complex, _ rhs: Complex) -> Complex {
  return Complex(lhs.real + rhs.real, lhs.imaginary + rhs.imaginary)
}
```

Reply:
I believe derived conformances already made stored properties differentiable.

Reply:
> Have you tried making real and imaginary differentiable?

I did in that when I mark a method as differentiable, I put the requirement that `@differentiable(vjp: _vjpFunc where T: Differentiable)`. I didn't require `real` and `imaginary` to be differentiable where I define the `Complex` struct, because when we are to differentiate the standard library's version of complex numbers, we won't be able to change it, and additionally it will restrict Complex numbers to what they can be more than what the standard library will likely have.

> It would be very interesting to add some tests that define a few operations in ways that autodiff can handle, and then see if autodiff computes the expected derivatives for those.

I'll give it a shot! 😄